### DISCUSSION TITLE
I can no longer pass empty string as argument values

### DISCUSSION THREAD

**Discussion by mattleibow** [cake-build/cake#3198] [Q&A] [answered]
### What You Are Seeing?

I can no longer run cake like this with an empty string:

```
dotnet cake --prefix=""
```

### What is Expected?

The argument in the cake string returns either null or an empty string:

```cs
var prefix = Argument("prefix", "");

// prefix == ""
```

### What version of Cake are you using?

1.0.0

### Are you running on a 32 or 64 bit system?

64

### What environment are you running on?  Windows? Linux? Mac?

Windows and macOS

### Are you running on a CI Server?  If so, which one?

DevOps and local

### How Did You Get This To Happen? (Steps to Reproduce)

No need for any special cake file, this is the tooling.

### Output Log

~~~sh
dotnet cake --prefix=""

Error: Expected an option value.

       --prefix=
               ^ Did you forget the option value?
~~~

**Comment by nils-a** [comment, marked as answer]
@mattleibow as said, this is a feature - or rather it's side [effect of Cake 1.0.0](https://cakebuild.net/docs/getting-started/upgrade#passing-empty-arguments) being able to accept flags in addition to key-value-arguments.

@devlead described it in [stack overflow](https://stackoverflow.com/a/66122190/180156)

> One key difference with Cake 1.0.0 is that it beyond keyvalue arguments (--key=value), supports flags (--flag), and multiple arguments with the same name (--key=value1 --key=value2). Which allows for much more flexibility than before.

As to you're question in gitter:

>But this is not quite the same, I am passing a string, and empty string.
Unless this is equivalent and I can do `--name ""`

No, that's not equivalent.
You'd have to supply a default in your Cake file. Like so:

```cakescript
var prefix = Argument<string>("prefix", "");
```

(Which then has the drawback that you probably will get the same value as if the argument was not set.)

**Comment by augustoproiete** [comment]
TL;DR; in Cake v1.0, Use a ` ` (space) instead of `=` if the value can be empty/null.

```
dotnet cake --prefix ""
```

> [_How do you pass empty arguments in Cake v1?_](https://stackoverflow.com/a/66125812/211672)

**Reply by gep13** [reply]
@augustoproiete @pascalberger I think that this is a good candidate for getting put into the docs "somewhere".  Thoughts?

**Reply by pascalberger** [reply]
> @augustoproiete @pascalberger I think that this is a good candidate for getting put into the docs "somewhere". Thoughts?

This is what I tried to document in [upgrade instructions](https://cakebuild.net/docs/getting-started/upgrade#passing-empty-arguments). But agree that it could be more clear there that it also applies for this case and additionally we should document this on the runner page. I can take this (https://github.com/cake-build/website/issues/1702).

**Reply by augustoproiete** [reply]
@gep13 Agreed it make sense to document further as @pascalberger's suggestion on https://github.com/cake-build/website/issues/1702

For a future version of Cake, it might be worth considering going a step further and update Cake _(1)_ to accept empty arguments and leave the validation to the user (and the `Argument<T>` alias already throws an exception if an argument doesn't have a default value/fallback).

---

_(1) Might require updates to Spectre.Console_

**Reply by RusKnyaz** [reply]
```sh
dotnet cake --prefix ""
```
It does not work if default value specified. For example:
```
var argCategory = Argument("category", "PureUnitTest");
```
I want to the argCategory be 'PureUnitTest' if the argument was not specified.
And I need the way to pass empty string to the 'category' argument.

But currently if I call the cake script using:
```
dotnet cake --category ""
```
I get the 'PureUnitTest' as category value.

**Reply by augustoproiete** [reply]
@RusKnyaz That's was a design choice with the thought that most people would want to get the default value in case the input is an empty string. For example, in cases where the value of the argument comes from an environment variable that might not exist, you could declare the command as:

```csharp
dotnet cake --category "%CATEGORY%"
```

Which would translate to an empty string if the environment variable is not present - and make the script use the default value.

We have the intent to add an [API that returns a dictionary with all the arguments provided in the command-line](https://github.com/cake-build/cake/issues/2690), which would allow you to check the presence of the argument and check if the value is an empty string. Would that be a good solution for what you're looking? PR's are welcome BTW.

Alternatively, if you have another idea on how to tackle that without breaking the existing behavior for everyone else, feel free to post on the [Ideas](https://github.com/cake-build/cake/discussions/categories/ideas) section.

**Reply by nils-a** [reply]
@RusKnyaz I have not tested it, but it should be possible to set the default to an empty string, and additionally test whether `--category` was passed in at all, using `HasArgument("category");`

So, (again, untested...)
```cakescript
var argCategory = Argument("category", "");
if(!HasArgument("category")) {
  argCategory = "PureUnitTest"
}
```

should probably fit your need.

**Reply by RusKnyaz** [reply]
> We have the intent to add an API that returns a dictionary with all the arguments provided in the command-line, which would allow you to check the presence of the argument and check if the value is an empty string. Would that be a good solution for what you're looking? 

Yes, I think it may resolve a lot of issues.

@nils-a thank you for suggestion. Of course I can solve the problem in a different ways. I just changed default value to empty string and adjusted CI configuration to the new arguments behaviour. 

But the root of the problem is how the empty string value is interpreted. As for me the empty string it is a value, and it is not the same as 'not specified' argument. And 'default' to be applied to non-specified arguments but not to empty strings. It seems logically. And it's not me, but that guy who want to replace empty string with default value should write something like:
```
var category = HasArgument("category") && argCategory.ToString() == string.Empty ? "PureUnitTest" : argCategory.ToString()
```

I am not calling for changing this now. Users should be polled before making breaking change.

BTW, I would vote for more descriptive way of arguments definition:
* specify available values for auto validation, something like: EnumArgument<BuildTargets>("target") or Argument("target").AvailableValues("build","pack","test");
* mark 'required parameter' like Argument().Required();
* specify description to be able show 'usage info' for script, like Argument(...).Describe("Specifies test categories to run"); 
In the same way you could specify default value for null and empty:
* Argument("category").NullValue("PureUnitTest").EmptyValue("");

**Reply by patriksvensson** [reply]
@augustoproiete I could add a fix for this in Spectre.Console so there wouldn't be a difference between empty parameters. After all, this is a bug.

**Reply by RusKnyaz** [reply]
BTW, Executing:
```c#
 CakeExecuteScript("external.cake", new CakeSettings{
       Arguments = new Dictionary<string, string> {
           {"myParam", ""},
        }
   });
```
Leads to the same exception.

**Reply by nick5454** [reply]
In Azure devops yml, I pass like so: Configuration "$(Configuration)" and it works fine

**Reply by nick5454** [reply]
But, what I noticed is that it's hard to find info on the migration and just help on 1.1.0. I think a page that's easy find b y googleing "cake build 1.1 migration". You get release notes, but not a migration guide hit. It would be nice if there was a document called "Cake 1.1 Migration guide" that was crawled by the spiders...

**Reply by augustoproiete** [reply]
@nick5454 Is this not helpful? https://cakebuild.net/docs/getting-started/upgrade

More specifically: https://cakebuild.net/docs/getting-started/upgrade#cake-cli-updates

**Reply by nick5454** [reply]
@augustoproiete I trying to convey that a person who does pipelines 20% of his duties doesn't have time to really search and find info.

So when googling, it's hard to find these answers. Maybe the pages aren't being hit by the spiders, but if that came up in a google search ( or duck duck go ) then there would be less questions and it would be easier for people to upgrade.
